[String] In-register comparison of small ASCII strings

milseman opened the pull request:
Adds a small-ASCII-string fast path to comparison that skips the spill to the stack. Adds it in inlinable code for a further speedup, but note that this might regress non-small-ASCII-strings. Still investigating.

<!-- What's in this pull request? -->
<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

milseman commented:
My preference is to keep it out of inlinable code, so trying that to compare. Locally, I saw some of the nano-benchmarks do less favorably such as Chars2.

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test

jrose-apple commented on stdlib/public/core/StringGuts.swift:
```diff
@@ -90,6 +90,11 @@ extension _StringGuts {
     @inline(__always) get { return _object.isSmall }
   }
 
+  @inlinable
```
This doesn't need to be `@inlinable` anymore, right?

milseman replied:
No, though everything inside it already is inlinable. I'll drop it

milseman commented:
Benchmark build at https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-perf/674/. I was getting some inconsistencies locally, and hope the not-hand-outlined version is just as good.

milseman committed:
[String] In-register smol ASCII string compare

Compare small strings in-register when they store ASCII (and thus NFC)
contents.

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test
